src/Shared/UrlDecoder/UrlDecoder.cs

```diff
@@ -345,5 +344,296 @@ private static bool SkipUnescape(int value1, int value2, bool isFormEncoding)
+        /// <param name="destinationIndex">The place to write to</param>
+        /// <param name="buffer">The char array</param>
+        private static bool DecodeCore(ref int sourceIndex, ref int destinationIndex, Span<char> buffer)
+        {
+            // preserves the original head. if the percent-encodings cannot be interpreted as sequence of UTF-8 octets,
+            // chars from this till the last scanned one will be copied to the memory pointed by writer.
+            var char1 = UnescapePercentEncoding(ref sourceIndex, buffer);
+            if (char1 == -1)
+            {
+                return false;
+            }
+
+            if (char1 == 0)
+            {
+                throw new InvalidOperationException("The path contains null characters.");
+            }
+
+            if (char1 <= 0x7F)
+            {
+                // first char < U+007f, it is a single char ASCII
+                buffer[destinationIndex++] = (char)char1;
+                return true;
+            }
+
+            // anticipate more chars
+            var currentDecodeBits = 0;
+            var charCount = 1;
+            var expectValueMin = 0;
+            if ((char1 & 0xE0) == 0xC0)
+            {
+                // 110x xxxx, expect one more char
+                currentDecodeBits = char1 & 0x1F;
+                charCount = 2;
+                expectValueMin = 0x80;
+            }
+            else if ((char1 & 0xF0) == 0xE0)
+            {
+                // 1110 xxxx, expect two more chars
+                currentDecodeBits = char1 & 0x0F;
+                charCount = 3;
+                expectValueMin = 0x800;
+            }
+            else if ((char1 & 0xF8) == 0xF0)
+            {
+                // 1111 0xxx, expect three more chars
+                currentDecodeBits = char1 & 0x07;
+                charCount = 4;
+                expectValueMin = 0x10000;
+            }
+            else
+            {
+                // invalid first char
+                return false;
+            }
+
+            var remainingChars = charCount - 1;
+            while (remainingChars > 0)
+            {
+                // read following three chars
+                if (sourceIndex == buffer.Length)
+                {
+                    return false;
+                }
+
+                var nextSourceIndex = sourceIndex;
+                var nextChar = UnescapePercentEncoding(ref nextSourceIndex, buffer);
+                if (nextChar == -1)
+                {
+                    return false;
+                }
+
+                if ((nextChar & 0xC0) != 0x80)
+                {
+                    // the follow up char is not in form of 10xx xxxx
+                    return false;
+                }
+
+                currentDecodeBits = (currentDecodeBits << 6) | (nextChar & 0x3F);
+                remainingChars--;
+
+                if (remainingChars == 1 && currentDecodeBits >= 0x360 && currentDecodeBits <= 0x37F)
```

Comment:
You shouldn't need this check or the "above 0x10FFFF?" check below. If you change line 540 below to call `Rune.TryCreate` instead of `new Rune(...)`, that will detect UTF-16 surrogate or above-0x10FFFF inputs and return _false_. So you'd end up with:

```cs
if (!Rune.TryCreate(currentDecodeBits, out Rune rune) || !rune.TryEncodeToUtf16(/* ... */))
{
    return false; // something went wrong
}
```

Reply:
Actually, I was considering this, and I was not sure about it. Thanks for the pointers.